hazelcast/src/hazelcast/client/connection/DefaultClientConnectionStrategy.cpp

```diff
@@ -0,0 +1,136 @@
+                            "DefaultClientConnectionStrategy::beforeGetConnection", "Client is offline.");
+                }
+            }
+
+            void DefaultClientConnectionStrategy::beforeOpenConnection(const Address &target) {
+                if (isClusterAvailable()) {
+                    return;
+                }
+                if (reconnectMode == config::ClientConnectionStrategyConfig::ASYNC && disconnectedFromCluster) {
+                    throw exception::HazelcastClientOfflineException(
+                            "DefaultClientConnectionStrategy::beforeGetConnection", "Client is offline");
+                }
+            }
+
+            void DefaultClientConnectionStrategy::onConnectToCluster() {
+            }
+
+            void DefaultClientConnectionStrategy::onDisconnectFromCluster() {
+                disconnectedFromCluster = true;
+                if (reconnectMode == config::ClientConnectionStrategyConfig::OFF) {
+                    shutdownWithExternalThread();
+                    return;
+                }
+                if (clientContext.getLifecycleService().isRunning()) {
+                    try {
+                        clientContext.getConnectionManager().connectToClusterAsync();
+                    } catch (exception::RejectedExecutionException &) {
+                        shutdownWithExternalThread();
+                    }
+                }
+            }
+
+            void DefaultClientConnectionStrategy::onConnect(const boost::shared_ptr<Connection> &connection) {
+            }
+
+            void DefaultClientConnectionStrategy::onDisconnect(const boost::shared_ptr<Connection> &connection) {
+            }
+
+            void DefaultClientConnectionStrategy::onHeartbeatStopped(const boost::shared_ptr<Connection> &connection) {
+
+                if (connection->isAuthenticatedAsOwner()) {
+                    boost::shared_ptr<exception::TargetDisconnectedException> exception = (
+                            exception::ExceptionBuilder<exception::TargetDisconnectedException>(
+                                    "DefaultClientConnectionStrategy::onHeartbeatStopped")
+                                    << "Heartbeat timed out to owner connection " << *connection).buildShared();
+                    connection->close("", exception);
+                }
+            }
+
+            void DefaultClientConnectionStrategy::onHeartbeatResumed(const boost::shared_ptr<Connection> &connection) {
+            }
+
+            void DefaultClientConnectionStrategy::shutdown() {
+            }
+
+            bool DefaultClientConnectionStrategy::isClusterAvailable() const {
+                return clientContext.getConnectionManager().getOwnerConnectionAddress().get() != NULL;
+            }
+
+            void DefaultClientConnectionStrategy::shutdownWithExternalThread() {
+                executor->execute(boost::shared_ptr<util::Runnable>(new ShutdownTask(clientContext)));
```

Comment:
This is done via a free external thread in java. 
The reason behind was shutdown logic. 
`shutdown` always waits for all services to stop. When it is run on executorService, it waits itself shutdown. 
In java there is 30 seconds time out there. And shutdown will always be 30 seconds in that case. 
That is why we used an external thread here.